kahuna/public/js/search-query/service.js

```diff
@@ -0,0 +1,51 @@
+import angular from 'angular';
+import Rx from 'rx';
+import '../util/rx';
+
+export const searchQueryService = angular.module('gr.searchQuery.service', ['util.rx']);
+
+searchQueryService.factory('searchQueryService', [function() {
+    // TODO: When we update rx, we need to swap the seed / acc around
+    // FIXME: We use startWith and set the start value as we want to emit
+    // the first value, but strangely that doesn't get saved into the state
+    const q$ = (query$, startWith = '') =>
+        query$.
+            scan(startWith, (state, fn) => fn(state).trim()).
+            distinctUntilChanged().
+            shareReplay(1).
+            startWith(startWith);
+
+    // TODO: It'd be nice to build this up from an API
+    const hasSpace    = s     => /\s/g.test(s);
+    const labelMatch  = label => new RegExp(`(label:|#)("|')?${label}(("|')|\\b)`, 'g');
+    const createLabel = label => hasSpace(label) ? `#"${label}"` : `#${label}`;
+
+    function hasLabel(q, label) {
+        return labelMatch(label).test(q);
+    }
+
+    function addLabel(query$) {
+        return label =>
+            query$.onNext(q => hasLabel(q, label) ? q : `${q} ${createLabel(label)}`);
+    }
+
+    function removeLabel(query$) {
+        return label => query$.onNext(q => q.replace(labelMatch(label), ''));
+    }
+
+    function setQuery(query$) {
+        return q => query$.onNext(() => q || '');
+    }
+
+    return q => {
+        const query$ = new Rx.Subject();
+        const service = {
+            q$:          q$(query$, q),
```

Comment:
What do we think about building `classes` this way?
Ot is perhaps using classes better?